Address issue #1495 by adding cypress spec name to mocha suite

egucciar opened the pull request:
Resolves: https://github.com/cypress-io/cypress/issues/1495

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cypress-io/cypress?pullRequest=3960) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cypress-io/cypress?pullRequest=3960) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cypress-io/cypress?pullRequest=3960) it.</sub>

egucciar commented:
I was able to see this worked for the Test Runner (in open), but i tried to verify this works in Run and It does not appear to be the case :( Will keep digging

egucciar commented:
I cant figure this out after 2 hours of going around the code, i guess i wont be splitting my tests in CircleCI by spec name / timings after all

egucciar commented:
![image](https://user-images.githubusercontent.com/2848185/56073290-53313e80-5d6f-11e9-9ae6-45680c4445fa.png)

It kind of works ? I wonder if CircleCI would accept this.

egucciar commented:
I wish i could just get it to be one level higher, hm..

egucciar commented:
@brian-mann i dont want to be much of a bother, but could anyone take a look at this PR / leave feedback / assist in updating these snapshot tests?

jennifer-shehane commented:
@egucciar It may be best to start by writing tests for the behavior your want - at least in this way, we can know what your end result should be and can offer more guidance. 

I'm also seeing the standard out printing this 'has file:' text in every single standard out now (captured in the failing snapshot tests), regardless of the use of a reporter. This is definitely not something we want printed in this way and I'm not sure where this text is coming from.

<img width="805" alt="Screen Shot 2019-04-16 at 10 23 05 PM" src="https://user-images.githubusercontent.com/1271364/56224818-47f35100-6096-11e9-9c06-aa8c7c8d8ded.png">

egucciar commented:
@jennifer-shehane thanks for the quick response. The `has file:` log comes from this commit from Chris Breiding 31 months ago:

https://github.com/cypress-io/cypress/commit/1d2383abe42a40c1ec0a7832d1de6e9f36b705e9

I shall comment it out.

I'm trying to run `reporter_spec.coffee` as this is where I'd add an additional assertion for the `file` i added (I am checking with CircleCI if this is sufficient - as the code stands today, adding the file somewhere in the report works but I'm unsure if it works for CircleCI if its not in a specific spot). As can be seen in the screenshot of the generated report, this 1 parent nesting, and 2 children testsuites, and the place where `file` is outputted is on a line with no test timing information so I'm feeling it wont work.

But, I can't get the test to run locally when doing this:

![image](https://user-images.githubusercontent.com/2848185/56233928-2f614780-6052-11e9-9825-373ebd52dfaf.png)

This is happening for me also on the _develop_ branch, after having built all the packages, and reinstalling the . node modules 

![image](https://user-images.githubusercontent.com/2848185/56234123-9979ec80-6052-11e9-9393-970da0d02fa3.png)


Unfortunately without a baseline of passing unit tests I feel uncomfortable updating the existing test with the new assertion just yet.

egucciar commented:
@jennifer-shehane thanks so much for the tip with the log, because that fixed all the tests!!! YAY!

But, I agree with your update that this is still a WIP because i believe the `file` should be on a Line that has the actual test information, i think i know what i can still try to do to fix this:

![image](https://user-images.githubusercontent.com/2848185/56235587-b5cb5880-6055-11e9-8bc9-808652ee277a.png)


But I'll have to look into it after our next quick launch, a few days. 

Basically, digging deeper into this LOC i added here to ensure it is added in the "right place" since i believe the reporter is just a pass-through and this only adds it on the one place where i added it:

![image](https://user-images.githubusercontent.com/2848185/56235675-dbf0f880-6055-11e9-82c3-901b4111320d.png)

thewalla07 commented:
@egucciar any update on this PR?

egucciar commented:
I don't have an update. I would really like this feature to be incorporated but i was more or less taking stabs in the dark and couldn't get it over the finish line. I don't recall what specifically blocked me.

gabbersepp commented:
@egucciar I am willing to finish this PR as I want to have the filename to be included in the test results. We use jenkins and a custom reporter and without the filename we have to to a full-text-search through all files to find the spec file that contains a failed test.

To do this it would be nice if you can tell me, what is currently missing and what is the current state? Reading through the comments does not help me unfortunatelly :-(

egucciar commented:
@gabbersepp Take a look at a mocha suite with timestamps generally used by CircleCI...the time  stamp has to  be present on the same tag that has the name of the file I think for CircleCI to correlate the time to the file name. 

I think my change may have propogated the  file name to the output.

egucciar commented:
@gabbersepp if you just  need a filename i  recommend taking my work, rebasing and seeing if it fits your needs. I needed the filename and timestamp to be in the same tag and couldn't figure out how to write test cases or get the code to do what I want because doing this very simple 2 line change took me hours.

egucciar commented:
also, i do remember what blocked me. couldn't get the tests to run, did not receive  any help or support on being able to run them. The comment where i showed the issue with the running test is where i left off. I was able to get the log as pointed out by jennifer to disappear, but the tests still would not run locally. Never did get a response to that comment and couldn't proceed.

gabbersepp commented:
thanks for your feedback @egucciar . Should be enough information for me :-)

gabbersepp commented:
@egucciar I cherry-picked your commits into a new branch: https://github.com/gabbersepp/cypress/tree/issue-1495
I also was able to access the `file` attribute within the reporter.

To fully understand your problem:
![image](https://user-images.githubusercontent.com/3234900/69674414-b83a6a00-109c-11ea-9964-7ed6eca4ff56.png)

You want the filename not only to be on the first `testsuite` tag but also on the second one?

egucciar commented:
Right, exactly, where the timing and test info are

gabbersepp commented:
I made some research but did not found any documentation that declares `file` to be part of the official junit xml format.
E.g.: https://github.com/junit-team/junit5/blob/master/platform-tests/src/test/resources/jenkins-junit.xsd
Unfortunatelly I am neither familiar with  circle ci nor with junit xml format.

Maybe the solution is to to use the filename as testsuite `name`? I took a quick look into the `jest-junit` project and they introduced an own flag for controling whether or not the filename should be mapped to the suite name.

egucciar commented:
It seems as though it was once supported in an old version of mocha junit. 

https://github.com/michaelleeallen/mocha-junit-reporter/commit/673aef95ebd56d90bb368f12ca3128b2c523d750


Honestly, this whole thing is confusing. Basically, the filename is the only was in which we can pass to Cypress what specs to run. Thus, split by filename would be the only split type we could leverage. Seems as though the expected behavior from circle would be similar to classname. Not sure where the mocha junit spec is documented or enforced.

https://github.com/circleci/circleci-docs/issues/1837

gabbersepp commented:
making this change in the `junit-reporter`:
![image](https://user-images.githubusercontent.com/3234900/69678245-592d2300-10a5-11ea-8328-98614cf8bc46.png)

Results in:
![image](https://user-images.githubusercontent.com/3234900/69678268-664a1200-10a5-11ea-8f22-5ae38572d65a.png)

So this change had to be made in the reporter project (https://github.com/michaelleeallen/mocha-junit-reporter) or you @egucciar write an own reporter that can do this.

gabbersepp commented:
> It seems as though it was once supported in an old version of mocha junit.
> 
> [michaelleeallen/mocha-junit-reporter@673aef9](https://github.com/michaelleeallen/mocha-junit-reporter/commit/673aef95ebd56d90bb368f12ca3128b2c523d750)
> 
> Honestly, this whole thing is confusing. Basically, the filename is the only was in which we can pass to Cypress what specs to run. Thus, split by filename would be the only split type we could leverage. Seems as though the expected behavior from circle would be similar to classname. Not sure where the mocha junit spec is documented or enforced.
> 
> [circleci/circleci-docs#1837](https://github.com/circleci/circleci-docs/issues/1837)

Please have a look at my comment. It was to the same time as yours so maybe you missed it. :-)
I think this old commit addresses another usecase as it puts the file attribute to the root suite I think.

egucciar commented:
Okay. I think I will also close this PR. I don't really have the capacity or need to work on this right now, but timings data would be interesting to me in the future. If it requires work elsewhere then it certainly doesn't need a PR here. My only question is if that change works by itself without any changes to Cypress ? Or alongside the change I made here ?

gabbersepp commented:
I do not know how you want to proceed here, but in my opinion your changes are the only one possible. Cypress can not do more as this seems to be a reporter problem.

So maybe you remove the `WIP` tag and wait if the Cypress Ci builds are green? I have not executed the tests locally but at least the last CI run was green. So maybe you can give it a try.

egucciar commented:
So you just think this is good to go on it's own?

gabbersepp commented:
> Okay. I think I will also close this PR. I don't really have the capacity or need to work on this right now, but timings data would be interesting to me in the future. If it requires work elsewhere then it certainly doesn't need a PR here. My only question is if that change works by itself without any changes to Cypress ? Or alongside the change I made here ?

Oh please do not close this PR :-). Just update your branch and remove the `WIP` tag. Your changes are absolutly necessary to make the file attribute available.
I had not executed the tests locally but maybe you can just update the branch, remove the WIP and wait if the cypress CI builds are green.

Otherwise I have to open a PR with your cherry picked commits^^

egucciar commented:
Alright well I'll need to merge develop in. Dunno why the builds aren't running either. Also I can remove the WIP tag. It accomplishes as much as it can I suppose!

gabbersepp commented:
> Alright well I'll need to merge develop in. Dunno why the builds aren't running either. Also I can remove the WIP tag. It accomplishes as much as it can I suppose!

Thanks! I assume that the remaining build chain stopped because of the WIP tag. But let's wait until the builds have finished.

gabbersepp commented:
All builds are gree. Looks good! 
Let me know if you need further assist.

jennifer-shehane commented on packages/server/lib/reporter.coffee:
```diff
@@ -40,8 +40,8 @@ getParentTitle = (runnable, titles) ->
-  if obj.file
-    console.log('has file:', obj.file)
+  # if obj.file
+    # console.log('has file:', obj.file)
```
Any reason this is left commented out? We do not advise leaving commented out code that is no longer in use or necessary. Why was this removed?

egucciar replied:
I can certainly remove them, I'm not sure why they were there to start. They added the logs to the circleci test you observed (see earlier comments), because once file was being passed in this logic was involved. Since I was unsure of the purpose of these lines of code, I commented them out, to gain an understanding of they were meant to be in there for some reason... If it's cool with you, I'll just remove the lines of code.

jennifer-shehane commented:
@egucciar Can you resolve conflicts and address this comment please? https://github.com/cypress-io/cypress/pull/3960#pullrequestreview-323472847 

If you are unable to work on this PR anymore, please close the PR.

egucciar commented:
Hi @jennifer-shehane , I don't know if you saw but I responded to your comment. Please let me know if removing the commented lines of code makes sense given my answer to your question.

egucciar commented:
@jennifer-shehane resolved conflicts  and removed the commented out lines of code assuming that's the right thing to do here.

egucciar commented:
Thanks @chrisbreiding for pushing this over the finish line 🙌

rtymchyk commented:
There was some conversation around making this compatible with CircleCI. It seems like the test results are now getting properly populated with filenames, and I can see it in the test reports. However in CircleCI I am still getting an error:

```
Error autodetecting timing type, falling back to weighting by name. Autodetect no matching filename or classname.  If file names are used, double check paths for absolute vs relative.
Example input file: "cypress/tests/contact.tests.js"
Example file from timings: ""
```

Just wondering if anyone had any luck with setting up parallelization in CircleCI with this test metadata?

jrnail23 commented:
@rtymchyk, I've tried it on my own, using their cli for test splitting, and the problem seems to be, at least in my case, that the timings data file (`/.circleci-task-data/circle-test-results/results.json`, by default) is missing file paths for each of my tests (`file: null`).

collinbachi commented:
We are also still having trouble getting CircleCI to map the filenames properly

egucciar commented:
Following @gabbersepp 's advice, one can modify the default mocha-junit-reporter to match CircleCI's expectations for XML. I have done this in the local project and may extrapolate this further to a forked repo and PR for mocha-junit-reporter, but will need to determine if that is a course of action I'll be taking. To assist others who are willing to do a similar fork process, this is what I ended up changing:

![image](https://user-images.githubusercontent.com/2848185/117896165-f249fc80-b28d-11eb-9f10-7f448e916474.png)

![image](https://user-images.githubusercontent.com/2848185/117896174-faa23780-b28d-11eb-9791-d3a59420e641.png)

I am able to see **partial CircleCI data** come through. **I am absolutely getting SOME timing data with SOME files now, at least, and pretty sure the rest may have to do with circle**. Circle requires green build for timing data so still working through why missing data for some of the specs...not sure if its related to the fact that with parallel jobs, it seems to only pick the result from "one" container. Hence I will work with CircleCI if this doesnt do the trick 100%, but im confident this does solve the issues on the XML reporting side.

* `testsuite` needs timing data
* I think - i could be wrong - the Root suite messes things.
